File: src/prototype/web/tests/scenario/counter_integration.test.ts
// Integration: attacker window firing → defender counter window opens →
// defender confirms → attacker window force-closes.
// Covers docs/design/input_system_defense_v1.md §D.1 – §D.2.

import { describe, expect, it } from "vitest";
import {
  initialActorState,
  initialGameState,
  stepSimulation,
  type GameState,
} from "../../src/state/game_state.js";
import type { InputFrame } from "../../src/input/types.js";
import type { Intent } from "../../src/input/intent.js";
import type { HandFSM } from "../../src/state/hand_fsm.js";
import type { FootFSM } from "../../src/state/foot_fsm.js";

function gripped(side: "L" | "R"): HandFSM {
  return Object.freeze({
    side,
    state: "GRIPPED" as const,
    target: "SLEEVE_R" as const,
    stateEnteredMs: 0,
    reachDurationMs: 0,
    lastParriedZone: null,
    lastParriedAtMs: Number.NEGATIVE_INFINITY,
  });
}
function foot(side: "L" | "R", state: FootFSM["state"]): FootFSM {
  return Object.freeze({ side, state, stateEnteredMs: 0 });
}

function frame(over: Partial<InputFrame> = {}): InputFrame {
  return Object.freeze({
    timestamp: 0,
    ls: { x: 0, y: 0 },
    rs: { x: 0, y: 0 },
    l_trigger: 0,
    r_trigger: 0,
    buttons: 0,
    button_edges: 0,
    device_kind: "Keyboard" as const,
    ...over,
  });
}

function intent(over: {
  hip?: Partial<Intent["hip"]>;
  grip?: Partial<Intent["grip"]>;
  discrete?: Intent["discrete"];
} = {}): Intent {
  return Object.freeze({
    hip: { hip_angle_target: 0, hip_push: 0, hip_lateral: 0, ...(over.hip ?? {}) },
    grip: {
      l_hand_target: null, l_grip_strength: 0, r_hand_target: null, r_grip_strength: 0,
      ...(over.grip ?? {}),
    },
    discrete: over.discrete ?? [],
  });
}

describe("counter window opens when attacker fires SCISSOR_SWEEP", () => {
  it("attacker OPENING same-frame causes counter OPENING with SCISSOR_COUNTER candidate", () => {
    // Seed conditions for SCISSOR_SWEEP (§8.2).
    const seed: GameState = Object.freeze({
      ...initialGameState(0),
      bottom: Object.freeze({
        ...initialActorState(0),
        leftFoot: foot("L", "LOCKED"),
        rightFoot: foot("R", "LOCKED"),
        leftHand: gripped("L"),
      }),
      top: Object.freeze({
        ...initialActorState(0),
        postureBreak: { x: 0.5, y: 0 },
      }),
    });
    const f = frame({ l_trigger: 0.8, ls: { x: 0.8, y: 0 } }); // +x sweep
    const i = intent({
      hip: { hip_angle_target: 0, hip_push: 0, hip_lateral: 0.8 },
      grip: { l_hand_target: "SLEEVE_R", l_grip_strength: 0.8 },
    });
    const res = stepSimulation(seed, f, i, {
      realDtMs: 16.67, gameDtMs: 16.67, confirmedTechnique: null,
    });
    expect(res.events.some((e) => e.kind === "WINDOW_OPENING")).toBe(true);
    expect(res.events.some((e) => e.kind === "COUNTER_WINDOW_OPENING")).toBe(true);
    expect(res.nextState.counterWindow.candidates).toContain("SCISSOR_COUNTER");
    // Lateral sign of attacker intent was +1.
    expect(res.nextState.attackerSweepLateralSign).toBe(1);
  });
});

describe("counter confirm force-closes the attacker window", () => {
  it("SCISSOR_COUNTER confirmed → attacker window flips to CLOSING", () => {
    // Build up the same seed; run enough frames to bring both windows to OPEN.
    let g: GameState = Object.freeze({
      ...initialGameState(0),
      bottom: Object.freeze({
        ...initialActorState(0),
        leftFoot: foot("L", "LOCKED"),
        rightFoot: foot("R", "LOCKED"),
        leftHand: gripped("L"),
      }),
      top: Object.freeze({
        ...initialActorState(0),
        postureBreak: { x: 0.5, y: 0 },
      }),
    });

    const baseFrame = (t: number) => frame({
      timestamp: t, l_trigger: 0.8, ls: { x: 0.8, y: 0 },
    });
    const baseIntent = intent({
      hip: { hip_angle_target: 0, hip_push: 0, hip_lateral: 0.8 },
      grip: { l_hand_target: "SLEEVE_R", l_grip_strength: 0.8 },
    });

    // Tick 1: windows OPENING.
    g = stepSimulation(g, baseFrame(0), baseIntent, {
      realDtMs: 16.67, gameDtMs: 16.67, confirmedTechnique: null,
    }).nextState;
    expect(g.judgmentWindow.state).toBe("OPENING");
    expect(g.counterWindow.state).toBe("OPENING");

    // Tick 2: both transition to OPEN (past 200ms).
    g = stepSimulation(g, baseFrame(250), baseIntent, {
      realDtMs: 16.67, gameDtMs: 16.67, confirmedTechnique: null,
    }).nextState;
    expect(g.judgmentWindow.state).toBe("OPEN");
    expect(g.counterWindow.state).toBe("OPEN");

    // Tick 3: defender commits SCISSOR_COUNTER — attacker window must close.
    const res = stepSimulation(g, baseFrame(300), baseIntent, {
      realDtMs: 16.67, gameDtMs: 16.67, confirmedTechnique: null,
      confirmedCounter: "SCISSOR_COUNTER",
    });
    expect(res.nextState.counterWindow.state).toBe("CLOSING");
    expect(res.nextState.judgmentWindow.state).toBe("CLOSING");
    expect(res.events.some((e) => e.kind === "COUNTER_CONFIRMED")).toBe(true);
  });

  it("TRIANGLE_EARLY_STACK confirmed clears top.arm_extracted on both sides", () => {
    // Seed a state where TRIANGLE would fire and arm_extracted was true.
    let g: GameState = Object.freeze({
      ...initialGameState(0),
      bottom: Object.freeze({
        ...initialActorState(0),
        leftFoot: foot("L", "UNLOCKED"),
        rightFoot: foot("R", "LOCKED"),
        leftHand: Object.freeze({
          ...gripped("L"), target: "COLLAR_R" as const,
        }),
      }),
      top: Object.freeze({
        ...initialActorState(0),
        armExtractedLeft: true,
      }),
      topArmExtracted: Object.freeze({
        left: true, right: false,
        leftSustainMs: 0, rightSustainMs: 0,
        leftSetAtMs: 0, rightSetAtMs: Number.NEGATIVE_INFINITY,
      }),
    });

    // Drive the window to OPEN. Because arm_extracted derives off hand+pull,
    // we also need a SLEEVE grip to keep it true across ticks.
    const baseFrame = (t: number) => frame({
      timestamp: t, l_trigger: 0.8, r_trigger: 0.9,
    });
    const baseIntent = intent({
      hip: { hip_angle_target: 0, hip_push: -0.6, hip_lateral: 0 },
      grip: {
        l_hand_target: "COLLAR_R", l_grip_strength: 0.8,
        r_hand_target: "SLEEVE_L", r_grip_strength: 0.9,
      },
    });
    // Seed a right hand gripped on SLEEVE_L too.
    g = Object.freeze({
      ...g,
      bottom: Object.freeze({
        ...g.bottom,
        rightHand: Object.freeze({
          ...gripped("R"), target: "SLEEVE_L" as const,
        }),
      }),
    });

    g = stepSimulation(g, baseFrame(0), baseIntent, {
      realDtMs: 16.67, gameDtMs: 16.67, confirmedTechnique: null,
    }).nextState;
    g = stepSimulation(g, baseFrame(250), baseIntent, {
      realDtMs: 16.67, gameDtMs: 16.67, confirmedTechnique: null,
    }).nextState;
    expect(g.counterWindow.state).toBe("OPEN");
    // arm_extracted remains true because we're still pulling.
    expect(g.top.armExtractedLeft).toBe(true);

    const res = stepSimulation(g, baseFrame(300), baseIntent, {
      realDtMs: 16.67, gameDtMs: 16.67, confirmedTechnique: null,
      confirmedCounter: "TRIANGLE_EARLY_STACK",
    });
    expect(res.nextState.top.armExtractedLeft).toBe(false);
    expect(res.nextState.top.armExtractedRight).toBe(false);
  });
});

File: src/prototype/unity/Assets/BJJSimulator/Tests/EditMode/Scenario/CounterIntegrationTest.cs
// NUnit EditMode mirror of src/prototype/web/tests/scenario/counter_integration.test.ts.
// Each [Test] corresponds to one it(...) case from the Stage 1 Vitest suite so a
// regression on either side produces a named, greppable failure.
//
// Integration: attacker window firing → defender counter window opens →
// defender confirms → attacker window force-closes.
// Covers docs/design/input_system_defense_v1.md §D.1 – §D.2.
//
// Run from Unity Editor: Window → General → Test Runner → EditMode.

using NUnit.Framework;

namespace BJJSimulator.Tests.Scenario
{
    // -------------------------------------------------------------------------
    // Helpers — mirror the gripped() / foot() / frame() / intent() factories in TS.
    // -------------------------------------------------------------------------

    public static class CounterIntegrationHelpers
    {
        public static HandFSM Gripped(HandSide side, GripZone target = GripZone.SleeveR) => new HandFSM
        {
            Side            = side,
            State           = HandState.Gripped,
            Target          = target,
            StateEnteredMs  = 0,
            ReachDurationMs = 0,
            LastParriedZone = GripZone.None,
            LastParriedAtMs = BJJConst.SentinelTimeMs,
        };

        public static FootFSM Foot(FootSide side, FootState state) => new FootFSM
        {
            Side           = side,
            State          = state,
            StateEnteredMs = 0,
        };

        // Equivalent to the `frame(over)` factory — start zeroed, then override.
        public static InputFrame Frame(
            long timestamp = 0,
            float lTrigger = 0f,
            float rTrigger = 0f,
            Vec2? ls = null,
            Vec2? rs = null) =>
            new InputFrame
            {
                Timestamp   = timestamp,
                Ls          = ls ?? Vec2.Zero,
                Rs          = rs ?? Vec2.Zero,
                LTrigger    = lTrigger,
                RTrigger    = rTrigger,
                Buttons     = ButtonBit.None,
                ButtonEdges = ButtonBit.None,
                DeviceKind  = DeviceKind.Keyboard,
            };

        public static Intent MakeIntent(
            HipIntent? hip = null,
            GripIntent? grip = null) =>
            new Intent
            {
                Hip      = hip  ?? HipIntent.Zero,
                Grip     = grip ?? GripIntent.Zero,
                Discrete = System.Array.Empty<DiscreteIntent>(),
            };

        public static StepOptions Opts(
            float dtMs = 16.67f,
            Technique? confirmedTechnique = null,
            CounterTechnique? confirmedCounter = null,
            DefenseIntent? defense = null) =>
            new StepOptions
            {
                RealDtMs           = dtMs,
                GameDtMs           = dtMs,
                ConfirmedTechnique = confirmedTechnique,
                ConfirmedCounter   = confirmedCounter,
                DefenseIntent      = defense,
            };

        public static bool HasEvent(SimEvent[] events, SimEventKind kind)
        {
            foreach (var e in events) if (e.Kind == kind) return true;
            return false;
        }

        public static bool ContainsCounter(CounterTechnique[] arr, CounterTechnique value)
        {
            if (arr == null) return false;
            foreach (var v in arr) if (v == value) return true;
            return false;
        }
    }

    // -------------------------------------------------------------------------
    // describe("counter window opens when attacker fires SCISSOR_SWEEP")
    // -------------------------------------------------------------------------

    [TestFixture]
    public class CounterWindowOpensOnScissorSweepTests
    {
        // it("attacker OPENING same-frame causes counter OPENING with SCISSOR_COUNTER candidate")
        [Test]
        public void AttackerOpeningSameFrameCausesCounterOpeningWithScissorCounterCandidate()
        {
            // Seed conditions for SCISSOR_SWEEP (§8.2).
            var seed = GameStateOps.InitialGameState(0);
            var bottom = seed.Bottom;
            bottom.LeftFoot  = CounterIntegrationHelpers.Foot(FootSide.L, FootState.Locked);
            bottom.RightFoot = CounterIntegrationHelpers.Foot(FootSide.R, FootState.Locked);
            bottom.LeftHand  = CounterIntegrationHelpers.Gripped(HandSide.L, GripZone.SleeveR);
            seed.Bottom = bottom;
            var top = seed.Top;
            top.PostureBreak = new Vec2(0.5f, 0f);
            seed.Top = top;

            var f = CounterIntegrationHelpers.Frame(
                lTrigger: 0.8f,
                ls:       new Vec2(0.8f, 0f)); // +x sweep
            var i = CounterIntegrationHelpers.MakeIntent(
                hip:  new HipIntent { HipAngleTarget = 0f, HipPush = 0f, HipLateral = 0.8f },
                grip: new GripIntent { LHandTarget = GripZone.SleeveR, LGripStrength = 0.8f, RHandTarget = GripZone.None, RGripStrength = 0f });

            var (next, events) = GameStateOps.Step(
                seed, f, i,
                CounterIntegrationHelpers.Opts(confirmedTechnique: null));

            Assert.IsTrue(CounterIntegrationHelpers.HasEvent(events, SimEventKind.WindowOpening),
                "expected WINDOW_OPENING event");
            Assert.IsTrue(CounterIntegrationHelpers.HasEvent(events, SimEventKind.CounterWindowOpening),
                "expected COUNTER_WINDOW_OPENING event");
            Assert.IsTrue(
                CounterIntegrationHelpers.ContainsCounter(next.CounterWindow.Candidates, CounterTechnique.ScissorCounter),
                "expected SCISSOR_COUNTER among counter candidates");
            // Lateral sign of attacker intent was +1.
            Assert.AreEqual(1, next.AttackerSweepLateralSign);
        }
    }

    // -------------------------------------------------------------------------
    // describe("counter confirm force-closes the attacker window")
    // -------------------------------------------------------------------------

    [TestFixture]
    public class CounterConfirmForceClosesAttackerWindowTests
    {
        // it("SCISSOR_COUNTER confirmed → attacker window flips to CLOSING")
        [Test]
        public void ScissorCounterConfirmedFlipsAttackerWindowToClosing()
        {
            // Build up the same seed; run enough frames to bring both windows to OPEN.
            var g = GameStateOps.InitialGameState(0);
            var bottom = g.Bottom;
            bottom.LeftFoot  = CounterIntegrationHelpers.Foot(FootSide.L, FootState.Locked);
            bottom.RightFoot = CounterIntegrationHelpers.Foot(FootSide.R, FootState.Locked);
            bottom.LeftHand  = CounterIntegrationHelpers.Gripped(HandSide.L, GripZone.SleeveR);
            g.Bottom = bottom;
            var top = g.Top;
            top.PostureBreak = new Vec2(0.5f, 0f);
            g.Top = top;

            InputFrame BaseFrame(long t) => CounterIntegrationHelpers.Frame(
                timestamp: t,
                lTrigger:  0.8f,
                ls:        new Vec2(0.8f, 0f));

            var baseIntent = CounterIntegrationHelpers.MakeIntent(
                hip:  new HipIntent { HipAngleTarget = 0f, HipPush = 0f, HipLateral = 0.8f },
                grip: new GripIntent { LHandTarget = GripZone.SleeveR, LGripStrength = 0.8f, RHandTarget = GripZone.None, RGripStrength = 0f });

            // Tick 1: windows OPENING.
            g = GameStateOps.Step(g, BaseFrame(0), baseIntent,
                CounterIntegrationHelpers.Opts(confirmedTechnique: null)).NextState;
            Assert.AreEqual(JudgmentWindowState.Opening, g.JudgmentWindow.State);
            Assert.AreEqual(CounterWindowState.Opening,  g.CounterWindow.State);

            // Tick 2: both transition to OPEN (past 200ms).
            g = GameStateOps.Step(g, BaseFrame(250), baseIntent,
                CounterIntegrationHelpers.Opts(confirmedTechnique: null)).NextState;
            Assert.AreEqual(JudgmentWindowState.Open, g.JudgmentWindow.State);
            Assert.AreEqual(CounterWindowState.Open,  g.CounterWindow.State);

            // Tick 3: defender commits SCISSOR_COUNTER — attacker window must close.
            var (next, events) = GameStateOps.Step(g, BaseFrame(300), baseIntent,
                CounterIntegrationHelpers.Opts(
                    confirmedTechnique: null,
                    confirmedCounter:   CounterTechnique.ScissorCounter));

            Assert.AreEqual(CounterWindowState.Closing,  next.CounterWindow.State);
            Assert.AreEqual(JudgmentWindowState.Closing, next.JudgmentWindow.State);
            Assert.IsTrue(CounterIntegrationHelpers.HasEvent(events, SimEventKind.CounterConfirmed),
                "expected COUNTER_CONFIRMED event");
        }

        // it("TRIANGLE_EARLY_STACK confirmed clears top.arm_extracted on both sides")
        [Test]
        public void TriangleEarlyStackConfirmedClearsTopArmExtractedOnBothSides()
        {
            // Seed a state where TRIANGLE would fire and arm_extracted was true.
            var g = GameStateOps.InitialGameState(0);
            var bottom = g.Bottom;
            bottom.LeftFoot  = CounterIntegrationHelpers.Foot(FootSide.L, FootState.Unlocked);
            bottom.RightFoot = CounterIntegrationHelpers.Foot(FootSide.R, FootState.Locked);
            bottom.LeftHand  = CounterIntegrationHelpers.Gripped(HandSide.L, GripZone.CollarR);
            // Seed a right hand gripped on SLEEVE_L too.
            bottom.RightHand = CounterIntegrationHelpers.Gripped(HandSide.R, GripZone.SleeveL);
            g.Bottom = bottom;

            var top = g.Top;
            top.ArmExtractedLeft = true;
            g.Top = top;

            g.TopArmExtracted = new ArmExtractedState
            {
                Left           = true,
                Right          = false,
                LeftSustainMs  = 0f,
                RightSustainMs = 0f,
                LeftSetAtMs    = 0,
                RightSetAtMs   = BJJConst.SentinelTimeMs,
            };

            // Drive the window to OPEN. Because arm_extracted derives off hand+pull,
            // we also need a SLEEVE grip to keep it true across ticks.
            InputFrame BaseFrame(long t) => CounterIntegrationHelpers.Frame(
                timestamp: t,
                lTrigger:  0.8f,
                rTrigger:  0.9f);

            var baseIntent = CounterIntegrationHelpers.MakeIntent(
                hip:  new HipIntent { HipAngleTarget = 0f, HipPush = -0.6f, HipLateral = 0f },
                grip: new GripIntent
                {
                    LHandTarget = GripZone.CollarR, LGripStrength = 0.8f,
                    RHandTarget = GripZone.SleeveL, RGripStrength = 0.9f,
                });

            g = GameStateOps.Step(g, BaseFrame(0), baseIntent,
                CounterIntegrationHelpers.Opts(confirmedTechnique: null)).NextState;
            g = GameStateOps.Step(g, BaseFrame(250), baseIntent,
                CounterIntegrationHelpers.Opts(confirmedTechnique: null)).NextState;

            Assert.AreEqual(CounterWindowState.Open, g.CounterWindow.State);
            // arm_extracted remains true because we're still pulling.
            Assert.IsTrue(g.Top.ArmExtractedLeft);

            var (next, _) = GameStateOps.Step(g, BaseFrame(300), baseIntent,
                CounterIntegrationHelpers.Opts(
                    confirmedTechnique: null,
                    confirmedCounter:   CounterTechnique.TriangleEarlyStack));

            Assert.IsFalse(next.Top.ArmExtractedLeft);
            Assert.IsFalse(next.Top.ArmExtractedRight);
        }
    }
}
